### ISSUE TITLE
i2pd specific attack?

### ISSUE THREAD

**Issue by diva-exchange** [PurpleI2P/i2pd#1915]
Running i2pd v2.47.0 in a docker container, this one:

https://hub.docker.com/layers/divax/i2p/current/images/sha256-eb5561394b32b799768293432e752c5de46d23efaed19cccbb025529471a6b8b?context=explore

Results in:

![image](https://user-images.githubusercontent.com/61228890/235357670-282bce98-0b3b-4c0f-9d77-1717340e5ec7.png)

Any feedback on this? Can diva.exchange help somehow to analyse the situation?

**Comment by Vort** [comment]
> Any feedback on this?

Network is under attack right now.

**Comment by twofaktor** [comment]
Bitcoin network affected too!

[imagen](https://user-images.githubusercontent.com/89636253/235375667-d7149d53-f250-461f-94f3-b98661de69f3.png)

**Comment by diva-exchange** [comment]
Note: i2pd v2.42.1 is not affected by the attack. One of the diva.exchange test networks is running on version v2.42.1 and it is doing fine.

**Comment by Vort** [comment]
Can you try to test 7c535159bc442b2ad539c4eb22eda00b018cc47d with disabled SSU2?
Not that I have much hopes about it, but why not.

**Comment by VictorBurtsev** [comment]
> Can you try to test [7c53515](https://github.com/PurpleI2P/i2pd/commit/7c535159bc442b2ad539c4eb22eda00b018cc47d) with disabled SSU2? Not that I have much hopes about it, but why not.

Hi, i`am tested it on i2pd:latest docker image with ssu2.enable=false

![Firefox_Screenshot_2023-05-01T07-32-02 319Z](https://user-images.githubusercontent.com/15078582/235423634-4b390028-3ac5-4a42-996e-743da2f04bea.png)

**Comment by diva-exchange** [comment]
> Hi, i`am tested it on i2pd:latest docker image with ssu2.enable=false

Tx! But: which docker image is it (link to hub.docker.com, if possible, please)? Are you sure that [7c53515](https://github.com/PurpleI2P/i2pd/commit/7c535159bc442b2ad539c4eb22eda00b018cc47d) is compiled in this image?

**Comment by VictorBurtsev** [comment]
> > Hi, i`am tested it on i2pd:latest docker image with ssu2.enable=false
> 
> Tx! But: which docker image is it (link to hub.docker.com, if possible, please)? Are you sure that [7c53515](https://github.com/PurpleI2P/i2pd/commit/7c535159bc442b2ad539c4eb22eda00b018cc47d) is compiled in this image?

Hi, tested on [this image](https://hub.docker.com/layers/purplei2p/i2pd/latest/images/sha256-ade0341e2c472b74bfe46d7340e163f291b9ec7aac4552bd7c1615ca82e4df34?context=explore)

**Comment by diva-exchange** [comment]
> Hi, tested on [this image](https://hub.docker.com/layers/purplei2p/i2pd/latest/images/sha256-ade0341e2c472b74bfe46d7340e163f291b9ec7aac4552bd7c1615ca82e4df34?context=explore)

Thanks a lot! From Timestamp-perspective, seems like the latest commit is included in this image. We'll do some compiles on older versions (which are tagged) within the next hours/days to get a better understanding which older versions work (like v.2.45...).

**Comment by SidorKozlov** [comment]
> Can you try to test [7c53515](https://github.com/PurpleI2P/i2pd/commit/7c535159bc442b2ad539c4eb22eda00b018cc47d) with disabled SSU2? Not that I have much hopes about it, but why not.

![2023-05-01T11:58:22,637658622+02:00](https://user-images.githubusercontent.com/132262654/235438028-5ff6e308-4bf1-4d0d-9dac-a1fa6ba7e285.png)

**Comment by SidorKozlov** [comment]
> Note: i2pd v2.42.1 is not affected by the attack. One of the diva.exchange test networks is running on version v2.42.1 and it is doing fine.

I can confirm 2.42.1 (SSU2 enabled again) works well even under high load (many SAM sessions).
![2023-05-01T12:30:58,302708521+02:00](https://user-images.githubusercontent.com/132262654/235441076-5db18bb0-72f5-495b-b962-8531d2291f3f.png)

**Comment by Vort** [comment]
Marker of successful attack is large amount of Floodfills in web console (> 2000).
Until this indicator goes away, problem cannot be said to be solved.

**Comment by schildbach** [comment]
My i2pd quit within the last 24h with error code 137 (out of memory), with this error message:

```
i2pd_1              | 15:11:10@352/error - SAM: Handshake read error: End of file
```

Is this a symptom of the attack?

**Comment by SidorKozlov** [comment]
> My i2pd quit within the last 24h with error code 137 (out of memory), with this error message:
> 
> ```
> i2pd_1              | 15:11:10@352/error - SAM: Handshake read error: End of file
> ```
> 
> Is this a symptom of the attack?

It is.

**Comment by Vort** [comment]
I saw 700 MB of RAM usage for my node because of attack.

**Comment by SidorKozlov** [comment]
Mine went well over 100k routers and floodfills on an 128GB machine, didn't check usage but killed it, because there was no point.

**Comment by SidorKozlov** [comment]
> Marker of successful attack is large amount of Floodfills in web console (> 2000). Until this indicator goes away, problem cannot be said to be solved.

I'm with you, it's not over @Vort
 
But I think it's worth to point out, that 0.42.1 resets router & floodfill count much more often, the numbers grow much slower with a lower ratio of floodfills.

Overall it is able to maintain my tunnels and allows me to be operational, while the newest versions (including the latest commit) die down very quickly,

**Comment by TpNYxtt11ox5TgZ** [comment]
Extending the length of the exploration tunnel and reducing the mass to 6 hop 1 tunnel also seems to mitigate the attack
But to be honest the best way is to add the java version of sybil analysis and block list function

**Comment by LLE8** [comment]
Debian 11.6 x86_64
RAM 1024 MB without swap

i2pd version 2.47.0-36-g4ebc7c97 (0.9.58)
Boost version 1.74.0
OpenSSL 1.1.1n  15 Mar 2022

kern.log
```
Apr 30 18:39:44 vds111111 kernel: [4829385.045811] Tasks state (memory values in pages):
Apr 30 18:39:44 vds111111 kernel: [4829385.045812] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
 ****
Apr 30 18:39:44 vds111111 kernel: [4829385.045834] [ 392881]   107 392881   368177   184528  1658880        0             0 i2pd
 ****
Apr 30 18:39:44 vds111111 kernel: [4829385.045924] Out of memory: Killed process 392881 (i2pd) total-vm:1472708kB, anon-rss:738112kB, file-rss:0kB, shmem-rss:0kB, UID:107 pgtables:1620kB oom_score_adj:0
```

**Comment by SidorKozlov** [comment]
> Debian 11.6 x86_64 RAM 1024 MB without swap
> 
> i2pd version 2.47.0-36-g4ebc7c97 (0.9.58) Boost version 1.74.0 OpenSSL 1.1.1n 15 Mar 2022
> 
> i2pd.conf
> 
> ```
> nettime.frompeers = false
> nat = false
> bandwidth = X
> notransit = false
> floodfill = true
> transittunnels = 16000
> ```
> 
> kern.log
> 
> ```
> Apr 30 18:39:44 vds111111 kernel: [4829385.045811] Tasks state (memory values in pages):
> Apr 30 18:39:44 vds111111 kernel: [4829385.045812] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
>  ****
> Apr 30 18:39:44 vds111111 kernel: [4829385.045834] [ 392881]   107 392881   368177   184528  1658880        0             0 i2pd
>  ****
> Apr 30 18:39:44 vds111111 kernel: [4829385.045924] Out of memory: Killed process 392881 (i2pd) total-vm:1472708kB, anon-rss:738112kB, file-rss:0kB, shmem-rss:0kB, UID:107 pgtables:1620kB oom_score_adj:0
> ```

This is expected considering the current situation.

**Comment by LLE8** [comment]
Less then 170 MB a few hours before

**Comment by SidorKozlov** [comment]
> Less then 170 MB a few hours before

Same here, my SAM app went from tons of peers to zero within minutes.

**Comment by SidorKozlov** [comment]
> Extending the length of the exploration tunnel and reducing the mass to 6 hop 1 tunnel also seems to mitigate the attack But to be honest the best way is to add the java version of sybil analysis and block list function

is the blacklist generated and held individually or is it shared?

**Comment by Vort** [comment]
For blocking to work correctly, attack origin needs to be uncovered first.
No one was able to do this yet as far as I know.
Of course, it is possible to block all proxies, all Tor nodes, but it's a dirty solution.

**Comment by twofaktor** [comment]
> 

Yes, I'm running 2.47.0 too. Yesterday I noticed this altercation and last night it seems that it was solved in some of my nodes, in others the error continues to appear, it seems that this problem has not yet been completely solved 😢

**Comment by twofaktor** [comment]
> > Bitcoin network was affected too!
> > ![imagen](https://user-images.githubusercontent.com/89636253/235375667-d7149d53-f250-461f-94f3-b98661de69f3.png)
> 
> @twofaktor I have been seeing the same with i2pd 2.47.0 for the past couple days.

Yes, I'm running the latest 2.47.0 too. Yesterday I noticed this altercation and last night it seems that it was solved in some of my nodes, in others the error continues to appear, and it seems that this problem has not yet been completely solved 😢
